### PR TITLE
Fix recursive CHMOD issue

### LINKED ISSUE #316: Recursive CHMOD issue
There are certain circumstances where Cache_Enabler_Disk::get_cache_dir() can return an empty string.

In this case, Cache_Enabler_Disk::get_cache_file() can return name with empty cache directory... eg: /https-index.html

Cache_Enabler_Disk::create_cache_file() will then attempt to call Cache_Enabler_Disk::mkdir_p() with "/" as the param.

This will cause Cache Enabler to try and recursively change permissions of all files under the root directory!

As well as causing all files in user home directory to change permissions, this can also crash PHP-FPM (as the pool socket is usually owned by the user).

The immediate fix is to ensure Cache_Enabler_Disk::get_cache_dir() ALWAYS returns at least the cache root path (CACHE_ENABLER_CACHE_DIR).

However, it may also be prudent to validate the $dir param in Cache_Enabler_Disk::mkdir_p() to ensure it contains the cache root path, and does not attempt parent directory traversal ("../")

### PR DESCRIPTION
**Pull request by robwoodgate**
Fixes #316

### CODE REVIEW

**Comment by robwoodgate** [comment]
@coreykn - this is a pretty urgent fix that prevents CE crashing websites (esp on PHP-FPM)

**Comment by futtta** [comment]
@svenba I've seen my site crashing as well when using keycdn on php-fpm. can't say if this was the root cause, but it seems likely?

**Comment by robwoodgate** [comment]
> @svenba I've seen my site crashing as well when using keycdn on php-fpm. can't say if this was the root cause, but it seems likely?

Took me many days of head-scratching and site outages to figure out... and still not entirely sure how it gets triggered, but after adding lots of error logging and debug_backtraces, I narrowed it down to the chain of events I detailed in #316. The PHP-FPM crash occurs because the permissions of the socket are changed by the recursive chmod.

**Comment by svenba** [comment]
Thank you for the pull request! I'm going to review this in the next couple of days

**Review by svenba — CHANGES_REQUESTED** [review]

File: inc/cache_enabler_disk.class.php [hunk 1]
```diff
@@ -602,7 +602,7 @@ private static function get_cache_dir( $url = null ) {
 
         $url_host = parse_url( $url, PHP_URL_HOST );
         if ( ! is_string( $url_host ) ) {
-            return '';
+            return CACHE_ENABLER_CACHE_DIR;
```
[the remark this answers is not in the document]
Comment (robwoodgate) [reply inc/cache_enabler_disk.class.php:605]: Hi @svenba - have you read through the related issue #316 ? This explains the reason why and how the result of this method is related to the recursive chmod.

In summary, the cache_dir is made up of three parts: CACHE_ENABLER_CACHE_DIR / $url_host  /  $url_path

It is not correct to return nothing if the url_host isn't specified. By returning CACHE_ENABLER_CACHE_DIR, at least the method is containing any subsequent actions to that folder.

In other words, it "fails safe" if $url_host isn't found.

**Review by svenba — APPROVED** [review]
